base/version_comparable_test.go

```diff
@@ -115,6 +115,115 @@ func TestInvalidComparableVersion(t *testing.T) {
 	}
 }
 
+func TestComparableVersionJSONRoundTrip(t *testing.T) {
+	json, err := JSONMarshal(ProductVersion)
+	require.NoError(t, err)
+	var version ComparableVersion
+	err = JSONUnmarshal(json, &version)
+	require.NoError(t, err)
+	require.True(t, ProductVersion.Equal(&version))
+	require.Equal(t, ProductVersion.String(), version.String())
+}
+
+func TestComparableVersionEmptyStringJSON(t *testing.T) {
+	var version ComparableVersion
+	err := JSONUnmarshal([]byte(`""`), &version)
+	require.NoError(t, err)
+	require.True(t, zeroComparableVersion().Equal(&version))
+	require.Equal(t, "0.0.0", zeroComparableVersion().String())
+	require.Equal(t, "0.0.0", version.String())
+}
+
+func TestAtLeastMinorDowngradeVersion(t *testing.T) {
+	testCases := []struct {
+		versionA       string
+		versionB       string
+		minorDowngrade bool
+	}{
+		{
+			versionA:       "1.0.0",
+			versionB:       "1.0.0",
+			minorDowngrade: false,
+		},
+		{
+			versionA:       "1.0.0",
+			versionB:       "2.0.0",
+			minorDowngrade: false,
+		},
+		{
+			versionA:       "2.0.0",
+			versionB:       "1.0.0",
+			minorDowngrade: true,
+		},
+		{
+			versionA:       "1.0.0",
+			versionB:       "1.0.1",
+			minorDowngrade: false,
+		},
+		{
+			versionA:       "1.0.1",
+			versionB:       "1.0.0",
+			minorDowngrade: false,
+		},
+		{
+			versionA:       "1.1.0",
+			versionB:       "1.0.0",
+			minorDowngrade: true,
+		},
+		{
+			versionA:       "1.0.0",
+			versionB:       "1.1.0",
+			minorDowngrade: false,
+		},
+		{
+			versionA:       "1.0.0",
+			versionB:       "1.0.0.1",
+			minorDowngrade: false,
+		},
+		{
+			versionA:       "1.0.0.1",
+			versionB:       "1.0.0",
+			minorDowngrade: false,
+		},
+		{
+			versionA:       "1.0.0.1",
+			versionB:       "1.0.0.2",
+			minorDowngrade: false,
+		},
+		{
+			versionA:       "1.0.0.2",
+			versionB:       "1.0.0.1",
+			minorDowngrade: false,
+		},
+		{
+			versionA:       "1.0.0-EE",
+			versionB:       "1.1.0-CE",
+			minorDowngrade: false,
+		},
+		{
+			versionA:       "2.2.0",
+			versionB:       "1.1.0",
+			minorDowngrade: true,
+		},
+		{
+			versionA:       "1.1.0",
+			versionB:       "2.2.0",
+			minorDowngrade: false,
```

Comment:
It looks like we still don't have a test case where a.major < b.major with a.minor > b.minor (e.g. 1.2.0 to 2.1.0), and vice versa (2.1.0 to 1.2.0).  Would be good to include for completeness.